api/konnect/v1alpha1/konnect_cloudgateway_dataplane_configuration_types.go

```diff
@@ -0,0 +1,257 @@
+}
+
+// ConfigurationDataPlaneGroupEnvironmentField specifies an environment variable field for the data-plane group.
+type ConfigurationDataPlaneGroupEnvironmentField struct {
+	// Name of the environment variable field to set for the data-plane group. Must be prefixed by KONG_.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="^KONG_."
+	Name string `json:"name"`
+	// Value assigned to the environment variable field for the data-plane group.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Required
+	Value string `json:"value"`
+}
+
+// ConfigurationDataPlaneGroupAutoscale specifies the autoscale configuration for the data-plane group.
+//
+// +kubebuilder:validation:XValidation:rule="self.type == 'ConfigurationDataPlaneGroupAutoscaleStatic' ? has(self.static) : true",message="static is required when type is ConfigurationDataPlaneGroupAutoscaleStatic"
+// +kubebuilder:validation:XValidation:rule="self.type == 'ConfigurationDataPlaneGroupAutoscaleAutopilot' ? has(self.autopilot) : true",message="autopilot is required when type is ConfigurationDataPlaneGroupAutoscaleAutopilot"
+type ConfigurationDataPlaneGroupAutoscale struct {
+	// ConfigurationDataPlaneGroupAutoscaleStatic specifies the static configuration for the data-plane group.
+	//
+	// +kubebuilder:validation:Optional
+	ConfigurationDataPlaneGroupAutoscaleStatic *ConfigurationDataPlaneGroupAutoscaleStatic `json:"static,omitempty"`
+
+	// ConfigurationDataPlaneGroupAutoscaleAutopilot specifies the autoscale configuration for the data-plane group.
+	//
+	// +kubebuilder:validation:Optional
+	ConfigurationDataPlaneGroupAutoscaleAutopilot *ConfigurationDataPlaneGroupAutoscaleAutopilot `json:"autopilot,omitempty"`
+
+	// Type of autoscaling to use.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=ConfigurationDataPlaneGroupAutoscaleStatic;ConfigurationDataPlaneGroupAutoscaleAutopilot
+	Type sdkkonnectcomp.ConfigurationDataPlaneGroupAutoscaleType `json:"type"`
+}
+
+// ConfigurationDataPlaneGroupAutoscaleAutopilot specifies the autoscale configuration for the data-plane group.
+type ConfigurationDataPlaneGroupAutoscaleAutopilot struct {
+	// Base number of requests per second that the deployment target should support.
+	//
+	// +kubebuilder:validation:Required
+	BaseRps int64 `json:"base_rps"`
+
+	// Max number of requests per second that the deployment target should support. If not set, this defaults to 10x base_rps.
+	MaxRps *int64 `json:"max_rps,omitempty"`
+}
+
+// ConfigurationDataPlaneGroupAutoscaleStatic specifies the static configuration for the data-plane group.
+type ConfigurationDataPlaneGroupAutoscaleStatic struct {
+	// Instance type name to indicate capacity.
+	//
+	// +kubebuilder:validation:Required
+	InstanceType sdkkonnectcomp.InstanceTypeName `json:"instance_type"`
```

Comment:
can we perform some kind of validation here? Do we know the available instance types upfront?

Reply:
I'd be inclined to add the enum here (based on https://github.com/Kong/sdk-konnect-go/blob/d511d983a17fb1be7ddd4950cc571f4482990127/openapi.yaml#L10414-L10417). What I'm worried about here is that the API might add new values and users will have to wait for us to release a new CRD version with that value added to the enum.

I'd suggest that we add a note about the current possible values (`small`, `medium` and `large`) and leave a link referring users to the API documentation e.g. https://docs.konghq.com/konnect/api/cloud-gateways/latest/#/Data-Plane%20Group%20Configurations/create-configuration. WDYT?

Reply:
sgtm

Reply:
👍 3aff77b6bf49a595643c5a2dff25638ce43c46a3